include/deal.II/numerics/vector_tools_project.templates.h

```diff
@@ -578,12 +570,14 @@ namespace VectorTools
       // copy vec into vec_result. we can't use vec_result itself above, since
       // it may be of another type than Vector<double> and that wouldn't
       // necessarily go together with the matrix and other functions
-      for (unsigned int i = 0; i < vec.size(); ++i)
-        ::dealii::internal::ElementAccess<VectorType>::set(vec(i),
-                                                           i,
+      for (const auto &it : dof.locally_owned_dofs())
+        ::dealii::internal::ElementAccess<VectorType>::set(vec(it),
+                                                           it,
```

Comment:
```suggestion
      for (unsigned int i = 0; i < vec.size(); ++i)
        ::dealii::internal::ElementAccess<VectorType>::set(vec(i),
                                                           i,
```
In the original code, you are using indices and iterate over all entries of the `vec` container. Here, you iterate over all locally owned dofs. I don't think this is correct, so I suggest to revert this change.

Reply:
You are right and the ``do_project`` function is called in serial cases only.